src/task_manager.cc

```diff
@@ -181,71 +202,95 @@ namespace detail {
+		}
+
+		epoch_for_new_tasks = epoch;
+	}
+
+	task_id task_manager::generate_horizon_task() {
 		// we are probably overzealous in locking here
+		task_id tid;
 		{
 			std::lock_guard lock(task_mutex);
+			tid = get_new_tid();
 			current_horizon_critical_path_length = max_pseudo_critical_path_length;
+			const auto previous_horizon = current_horizon;
+			current_horizon = reduce_execution_front(task::make_horizon_task(tid));
+			if(previous_horizon) { set_epoch_for_new_tasks(*previous_horizon); }
+		}
 
-			auto* previous_horizon_task = current_horizon_task;
-			current_horizon_task = &register_task_internal(task::make_horizon_task(get_new_tid()));
-
-			// add dependencies from a copy of the front to this task
-			auto current_front = get_execution_front();
-			for(task* front_task : current_front) {
-				if(front_task != current_horizon_task) { add_dependency(current_horizon_task, front_task); }
-			}
-
-			// apply the previous horizon to buffers_last_writers and last_collective_tasks data structs
-			if(previous_horizon_task != nullptr) {
-				const task_id prev_hid = previous_horizon_task->get_id();
-				for(auto& [_, buffer_region_map] : buffers_last_writers) {
-					buffer_region_map.apply_to_values([prev_hid](std::optional<task_id> tid) -> std::optional<task_id> {
-						if(!tid) return tid;
-						return {std::max(prev_hid, *tid)};
-					});
-				}
-				for(auto& [cgid, tid] : last_collective_tasks) {
-					tid = std::max(prev_hid, tid);
-				}
-				for(auto& [hoid, tid] : host_object_last_effects) {
-					tid = std::max(prev_hid, tid);
-				}
+		// it's important that we don't hold the lock while doing this
+		invoke_callbacks(tid);
+		return tid;
+	}
 
-				// We also use the previous horizon as the new init task for host-initialized buffers
-				current_init_task_id = prev_hid;
-			}
+	task_id task_manager::generate_epoch_task(epoch_action action) {
+		// we are probably overzealous in locking here
+		task_id tid;
+		{
+			std::lock_guard lock(task_mutex);
+			tid = get_new_tid();
+			const auto new_epoch = reduce_execution_front(task::make_epoch(tid, action));
+			compute_dependencies(new_epoch);
+			set_epoch_for_new_tasks(new_epoch);
+			current_horizon = std::nullopt; // this horizon is now behind the epoch_for_new_tasks, so it will never become an epoch itself
+			current_horizon_critical_path_length = max_pseudo_critical_path_length; // the explicit epoch resets the need to create horizons
 		}
 
 		// it's important that we don't hold the lock while doing this
-		invoke_callbacks(current_horizon_task->get_id(), task_type::HORIZON);
+		invoke_callbacks(tid);
+		return tid;
 	}
 
-	void task_manager::clean_up_pre_horizon_tasks() {
-		task_id deletion_task_id = horizon_task_id_for_deletion.exchange(nothing_to_delete);
-		if(deletion_task_id != nothing_to_delete) {
+	void task_manager::prune_tasks_before_last_epoch_reached() {
+		const auto prune_before = latest_epoch_reached.get();
+		if(prune_before > last_pruned_before) {
```

Comment:
Not sure I agree. The `>` carries the semantics that the last-pruning-epoch has advanced, rather than just changed. Since `last_pruned_before` is only accessed inside this function, I'd argue this is pretty traceable.

Reply:
Fair enough.